File: 0112-path-sum/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool hasPathSum(TreeNode* root, int targetSum) {
        if (!root) return false; // Si el nodo es nulo, no hay camino.

        // Restamos el valor del nodo actual del targetSum
        targetSum -= root->val;
        
        // Si llegamos a una hoja y targetSum es 0, encontramos un camino válido.
        if (!root->left && !root->right) return targetSum == 0;
        
        // Recursión: Si cualquiera de los dos caminos devuelve true, retornamos true.
        return hasPathSum(root->left, targetSum) || hasPathSum(root->right, targetSum);
    }
};

File: 0112-path-sum/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int i = 0;
    public boolean hasPathSum(TreeNode root, int targetSum) {
        if(root==null) return false;
       
        if(targetSum - root.val == 0 && i != 0) {
            if(root.left == null && root.right == null){
                return true;
            }
        }
        i++;
        if(hasPathSum(root.left,targetSum - root.val)) return true;
        
        if(hasPathSum(root.right,targetSum - root.val)) return true;
        
        else {
            if(targetSum == 0) return false;
            if(root.left == null && root.right == null && i == 1) return true;
        }

        return false;
    }
}
